File: mparticle/ios/Plugin/MparticlePlugin.swift
import Foundation
import Capacitor
import mParticle_Apple_SDK
import AppTrackingTransparency

public typealias JSObject = [String: Any]
public typealias JSArray = [JSObject]

@objc(MparticlePlugin)
public class MparticlePlugin: CAPPlugin {
    private let implementation = Mparticle()
    public let LOG_TAG = "bitburst.mparticle.plugin "

    override public func load() {
        // for iOS we need to delay initializing for ATT status response
        // initial load moved to init()
    }

    @objc func `init`(_ call: CAPPluginCall) {
        implementation.start(getConfig())
        call.resolve()
    }

    @objc func identifyUser(_ call: CAPPluginCall) {
        call.keepAlive = true
        let email = call.getString("email", "not set")
        let customerId = call.getString("customerId", "not set")
        let other = call.getString("other", "not set")

        MParticle.sharedInstance().identity.identify(implementation.identityRequest(email, customerId, other)!, completion: {(identityResult: MPIdentityApiResult?, _: Error?) in
            if identityResult?.user.userId != nil {
                call.resolve([
                    "userId": identityResult?.user.userId ?? ""
                ])
            } else {
                print(self.LOG_TAG + "failed to identify user.")
                call.reject("failed to identify user.")
            }
        })
    }

    @objc func setUserAttribute(_ call: CAPPluginCall) {
        let userId = call.getString("userId") ?? ""
        let name = call.getString("name") ?? ""
        let value = call.getString("value") ?? ""
        implementation.setUserAttribute(userId, name, value)
        call.resolve()
    }

    @objc func setUserAttributes(_ call: CAPPluginCall) {
        let userId = call.getString("userId") ?? ""
        let attributes = (call.getArray("attributes") ?? []) as! JSArray
        implementation.setUserAttributes(userId, attributes)
        call.resolve()
    }

    @objc func setGDPRConsent(_ call: CAPPluginCall) {
        let analytics = call.getObject("analytics") ?? [:]
        let advertising = call.getObject("advertising") ?? [:]
        let general = call.getObject("general") ?? [:]
        implementation.addGDPRConsentState(analytics, advertising, general)
        call.resolve()
    }

    @objc func getGDPRConsent(_ call: CAPPluginCall) {
        let consents = implementation.getGDPRConsent()
        call.resolve([
            "consents": consents!
        ])
    }

    @objc func getMPID(_ call: CAPPluginCall) {
        let mpid = implementation.currentUser()!.userId.stringValue
        call.resolve([
            "userId": mpid
        ])
    }

    @objc func trackEvent(_ call: CAPPluginCall) {
        let name = call.getString("name") ?? ""
        let data = call.getObject("data") ?? [:]
        implementation.trackEvent(name, data)
        call.resolve()
    }

    @objc func trackPageView(_ call: CAPPluginCall) {
        let name = call.getString("name") ?? ""
        let data = call.getObject("data") ?? [:]
        implementation.trackPageView(name, data)
        call.resolve()
    }

    @objc func trackPurchase(_ call: CAPPluginCall) {
        let name = call.getString("name", "not set")
        let sku = call.getString("sku", "not set")
        let quantity = call.getInt("quantity", -1)
        let price = call.getFloat("price", -1)
        let transactionId = call.getString("transactionId", "not set")
        let customAttributes = call.getObject("customAttributes")

        implementation.trackPurchaseEvent(name, sku, quantity, price, transactionId, customAttributes)
        call.resolve()
    }
}

File: mparticle/android/src/main/java/net/bitburst/plugins/mparticle/MparticlePlugin.java
package net.bitburst.plugins.mparticle;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.mparticle.consent.GDPRConsent;
import com.mparticle.identity.IdentityApiRequest;
import com.mparticle.identity.MParticleUser;
import java.text.ParseException;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import org.json.JSONException;

@CapacitorPlugin(name = "Mparticle")
public class MparticlePlugin extends Plugin {

    public static final String LOG_TAG = "bitburst.mparticle.plugin ";
    private static Mparticle implementation = null;

    @Override
    public void load() {
        implementation = new Mparticle(this);
        super.load();
    }

    @PluginMethod
    public void init(PluginCall call) {
        call.resolve();
    }

    @PluginMethod
    public void identifyUser(PluginCall call) {
        JSObject data = call.getData();
        try {
            IdentityApiRequest request = implementation.identityRequest(call, data);
            Mparticle
                .sharedInstance()
                .Identity()
                .identify(request)
                .addFailureListener(identityHttpResponse -> call.reject(LOG_TAG, Objects.requireNonNull(identityHttpResponse).toString()))
                .addSuccessListener(
                    identityApiResult -> {
                        if (data != null) {
                            Iterator<String> iterator = data.keys();
                            while (iterator.hasNext()) {
                                String key = iterator.next();
                                try {
                                    MParticleUser user = identityApiResult.getUser();
                                    user.setUserAttribute(key, data.get(key));

                                    JSObject ret = new JSObject();
                                    ret.put("userId", Long.toString(user.getId()));
                                    call.resolve();
                                } catch (JSONException e) {
                                    e.printStackTrace();
                                    call.reject(LOG_TAG, e.getLocalizedMessage());
                                }
                            }
                        }
                    }
                );
        } catch (JSONException e) {
            e.printStackTrace();
            call.reject(LOG_TAG, e.getLocalizedMessage());
        }
    }

    @PluginMethod
    public void setUserAttribute(PluginCall call) {
        String userId = call.getString("userId");
        String name = call.getString("name");
        JSObject value = call.getObject("value");
        var success = implementation.setUserAttribute(userId, name, value);
        if (!success) {
            call.reject(LOG_TAG, "failed to set user attribute");
            return;
        }
        call.resolve();
    }

    @PluginMethod
    public void setUserAttributes(PluginCall call) {
        try {
            List<Object> attributes = call.getArray("attributes").toList();
            String userId = call.getString("userId");
            var success = implementation.setUserAttributes(userId, attributes);
            if (!success) {
                call.reject(LOG_TAG, "failed to set user attributes");
                return;
            }
            call.resolve();
        } catch (JSONException e) {
            e.printStackTrace();
            call.reject(LOG_TAG, e.getLocalizedMessage());
        }
    }

    @PluginMethod
    public void setGDPRConsent(PluginCall call) {
        try {
            implementation.addGDPRConsentState(call.getData());
            call.resolve();
        } catch (JSONException | ParseException e) {
            e.printStackTrace();
            call.reject(LOG_TAG, e.getLocalizedMessage());
        }
    }

    @PluginMethod
    public void getGDPRConsent(PluginCall call) {
        try {
            Map<String, GDPRConsent> consents = implementation.getGDPRConsent();
            JSObject ret = new JSObject().put("consents", consents);
            call.resolve(ret);
        } catch (JSONException e) {
            e.printStackTrace();
            call.reject(LOG_TAG, e.getLocalizedMessage());
        }
    }

    @PluginMethod
    public void getMPID(PluginCall call) {
        long mpid = implementation.currentUser().getId();
        JSObject ret = new JSObject().put("userId", Long.toString(mpid));
        call.resolve(ret);
    }

    @PluginMethod
    public void trackEvent(PluginCall call) {
        JSObject callData = call.getData();
        if (callData == null) return;
        try {
            Mparticle.sharedInstance().logEvent(implementation.trackEvent(callData));
        } catch (JSONException e) {
            e.printStackTrace();
            call.reject(LOG_TAG, "failed to track event ", e);
        }
        call.resolve();
    }

    @PluginMethod
    public void trackPageView(PluginCall call) {
        JSObject callData = call.getData();
        if (callData == null) return;
        try {
            Mparticle.sharedInstance().logScreen(implementation.trackEvent(callData));
        } catch (JSONException e) {
            e.printStackTrace();
            call.reject(LOG_TAG, "failed to track page view ", e);
        }
        call.resolve();
    }

    @PluginMethod
    public void trackPurchase(PluginCall call) {
        JSObject callData = call.getData();
        if (callData == null) return;
        try {
            Mparticle.sharedInstance().logEvent(implementation.trackPurchaseEvent(callData));
        } catch (JSONException e) {
            e.printStackTrace();
            call.reject(LOG_TAG, "failed to track purchase event ", e);
        }
        call.resolve();
    }
}
